In [1]:
import pandas as pd

data = pd.read_csv('Data/mlb_elo.csv')
data.shape

(220946, 26)

There are 220946 rows in the data, I first tired to drop all Nan values, however, data reduced to 1580 rows after drop all Nan value, it is too small to prediction. And I noticed that most Nan value are from feature 'playoff', so I removed column 'playoff', then drop all other Nan values. 

In [2]:
data = data.drop('playoff', axis=1)
data = data.dropna()
data

date  season  neutral team1 team2     elo1_pre     elo2_pre  \
20      2020-10-02    2020        0   SDP   STL  1522.312734  1520.100045   
21      2020-10-02    2020        0   CHC   FLA  1517.011886  1478.403936   
22      2020-10-01    2020        0   LAD   MIL  1596.678175  1502.693145   
23      2020-10-01    2020        0   SDP   STL  1519.722828  1522.689951   
24      2020-10-01    2020        0   OAK   CHW  1545.674260  1507.785767   
...            ...     ...      ...   ...   ...          ...          ...   
186526  1913-04-10    1913        0   PHI   LAD  1501.215000  1476.395000   
186527  1913-04-10    1913        0   MIN   NYY  1526.658000  1458.322000   
186528  1913-04-10    1913        0   BOS   OAK  1560.562000  1540.280000   
186529  1913-04-10    1913        0   BAL   DET  1455.314000  1487.913000   
186530  1913-04-09    1913        0   LAD   PHI  1477.842000  1499.767000   

        elo_prob1  elo_prob2    elo1_post  ...  pitcher1_rgs  pitcher2_rgs  \
20       0.550128   0.449872  1525.832831  ...     47.400000     57.310588   
21       0.617874   0.382126  1513.409248  ...     60.622663     50.317526   
22       0.712092   0.287908  1598.290926  ...     56.674581     57.809816   
23       0.540271   0.459729  1522.312734  ...     52.966274     52.749116   
24       0.616569   0.383431  1547.618027  ...     50.478180     49.661565   
...           ...        ...          ...  ...           ...           ...   
186526   0.569799   0.430201  1496.632000  ...     49.837000     51.308000   
186527   0.629840   0.370160  1527.637000  ...     47.400000     47.400000   
186528   0.563384   0.436616  1558.977000  ...     47.308000     51.478000   
186529   0.487628   0.512372  1457.238000  ...     47.400000     50.759000   
186530   0.502986   0.497014  1476.395000  ...     59.720000     47.400000   

        pitcher1_adj pitcher2_adj rating_prob1  rating_prob2  rating1_post  \
20        -17.746023    29.444071     0.461857      0.538143   1534.038131   
21         40.641459     5.064633     0.673112      0.326888   1517.959644   
22         21.716845    33.175306     0.673916      0.326084   1606.065971   
23          7.282398     7.517466     0.544395      0.455605   1530.728125   
24         -1.921507    -6.888932     0.558056      0.441944   1532.981409   
...              ...          ...          ...           ...           ...   
186526      5.671534    12.777753     0.570321      0.429679   1496.875000   
186527      6.159533     3.262440     0.634058      0.365942   1527.618000   
186528      7.250708    11.261942     0.569341      0.430659   1559.060000   
186529      8.374392     3.151146     0.490798      0.509202   1457.304000   
186530     12.777753     5.671534     0.509368      0.490632   1476.113000   

        rating2_post  score1  score2  
20       1509.745224     4.0     0.0  
21       1475.108982     0.0     2.0  
22       1506.703444     3.0     0.0  
23       1513.055229    11.0     9.0  
24       1513.624593     6.0     4.0  
...              ...     ...     ...  
186526   1480.541000     3.0    11.0  
186527   1457.312000     2.0     1.0  
186528   1541.910000     9.0    10.0  
186529   1485.946000     3.0     1.0  
186530   1501.303000     0.0     1.0  

[186502 rows x 25 columns]

In this datasets, there are 4 columns in strings, 'team1', 'team2', 'pitcher1', 'pitcher2'. I then encoded this 4 columns by using LabelEncoder from sklearn. And delete a column 'date' which is not used.

In [3]:
from sklearn import preprocessing
team_encodingColumns = data['team1'].append(data['team2'])
pitcher_encodingColumns = data['pitcher1'].append(data['pitcher2'])

team_encoding = preprocessing.LabelEncoder()
pitcher_encoding = preprocessing.LabelEncoder()
team_encoding.fit(team_encodingColumns)
pitcher_encoding.fit(pitcher_encodingColumns)

data['team1'] = team_encoding.transform(data['team1'])
data['team2'] = team_encoding.transform(data['team2'])
data['pitcher1'] = pitcher_encoding.transform(data['pitcher1'])
data['pitcher2'] = pitcher_encoding.transform(data['pitcher2'])

data = data.drop('date',axis=1)
data.head()

season  neutral  team1  team2     elo1_pre     elo2_pre  elo_prob1  \
20    2020        0     22     25  1522.312734  1520.100045   0.550128   
21    2020        0      5     11  1517.011886  1478.403936   0.617874   
22    2020        0     14     15  1596.678175  1502.693145   0.712092   
23    2020        0     22     25  1519.722828  1522.689951   0.540271   
24    2020        0     19      6  1545.674260  1507.785767   0.616569   

    elo_prob2    elo1_post    elo2_post  ...  pitcher1_rgs  pitcher2_rgs  \
20   0.449872  1525.832831  1516.579948  ...     47.400000     57.310588   
21   0.382126  1513.409248  1482.006574  ...     60.622663     50.317526   
22   0.287908  1598.290926  1501.080394  ...     56.674581     57.809816   
23   0.459729  1522.312734  1520.100045  ...     52.966274     52.749116   
24   0.383431  1547.618027  1505.842000  ...     50.478180     49.661565   

    pitcher1_adj  pitcher2_adj  rating_prob1  rating_prob2  rating1_post  \
20    -17.746023     29.444071      0.461857      0.538143   1534.038131   
21     40.641459      5.064633      0.673112      0.326888   1517.959644   
22     21.716845     33.175306      0.673916      0.326084   1606.065971   
23      7.282398      7.517466      0.544395      0.455605   1530.728125   
24     -1.921507     -6.888932      0.558056      0.441944   1532.981409   

    rating2_post  score1  score2  
20   1509.745224     4.0     0.0  
21   1475.108982     0.0     2.0  
22   1506.703444     3.0     0.0  
23   1513.055229    11.0     9.0  
24   1513.624593     6.0     4.0  

[5 rows x 24 columns]

First split the data to training data and testing data with seperate sample and labels. Labels are two scores. 

In [4]:
from sklearn.model_selection import train_test_split
label = data[['score1','score2']]
sample = data.drop(['score1','score2'],axis=1)
sample_train, sample_test, label_train, label_test = train_test_split(sample, label, test_size=0.1)

I first tried the Linear Regression model to make the prediction.

In [5]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

LR = LinearRegression()
LR.fit(sample_train,label_train)
LR_predict = LR.predict(sample_test)
accuracy_LR = r2_score(label_test,LR_predict)
accuracy_LR

0.42002859849050417

The accuracy of this model is very low, since there is no parameters we can change to improve the accuracy, so then I tried the Random Forest Regression model.

In [6]:
from sklearn.ensemble import RandomForestRegressor

RF = RandomForestRegressor(n_estimators=10)
RF.fit(sample_train,label_train)
RF_predict = RF.predict(sample_test)
accuracy_RF = r2_score(label_test,RF_predict)
accuracy_RF

0.45752623481706

The accuracy for this model is also very low, then I tried to change some of the parameters to improve the accuracy. 

At first, I planed to try to change the maximun depth of random tree. Since when the tree keep expand untill all leaves very pure, then it may have a case that have some leaves with only few samples, and that may cause overfitting. However, I still cannot find a way to figure out what is the final depth when the leaves are all pure, so I did not know which range of depth I can choose. In another way, when the depth of the tree is decreased, even overfitting does not happen, the tree may not pure enough to make better prediction accuracy. 

Then I tried to add more random trees to make the model more complex.

In [7]:
RF_1 = RandomForestRegressor(n_estimators=20)
RF_1.fit(sample_train,label_train)
RF_predict_1 = RF_1.predict(sample_test)
accuracy_RF1 = r2_score(label_test,RF_predict_1)
accuracy_RF1

0.48972322882807356

In [8]:
RF_2 = RandomForestRegressor(n_estimators=50)
RF_2.fit(sample_train,label_train)
RF_predict_2 = RF_2.predict(sample_test)
accuracy_RF2 = r2_score(label_test,RF_predict_2)
accuracy_RF2

0.5054745144075841

In [9]:
RF_3 = RandomForestRegressor(n_estimators=100)
RF_3.fit(sample_train,label_train)
RF_predict_3 = RF_3.predict(sample_test)
accuracy_RF3 = r2_score(label_test,RF_predict_3)
accuracy_RF3

0.5097448208974248

In [10]:
RF_4 = RandomForestRegressor(n_estimators=200)
RF_4.fit(sample_train,label_train)
RF_predict_4 = RF_4.predict(sample_test)
accuracy_RF4 = r2_score(label_test,RF_predict_4)
accuracy_RF4

0.5131180509025598

As the result, the accuracy will increasing while more random tree is builded, but the implementation would be very slow and the accuracy only have few improve. 

Then I also tried with Gradient Boost and Neural Network meodel

The Gradient Boost only take one column of label value, so I make two model for each scores. 

In [11]:
from sklearn.ensemble import GradientBoostingRegressor

GB1 = GradientBoostingRegressor()
GB1.fit(sample_train,label_train['score1'])
GB_predict1 = GB1.predict(sample_test)
accuracy_GB_1 = r2_score(label_test['score1'],GB_predict1)
print(accuracy_GB_1)

GB2 = GradientBoostingRegressor()
GB2.fit(sample_train,label_train['score2'])
GB_predict2 = GB2.predict(sample_test)
accuracy_GB_2 = r2_score(label_test['score2'],GB_predict2)
print(accuracy_GB_2)


0.20820976532088775
0.2577883155497108


In [12]:
from sklearn.neural_network import MLPRegressor

MLP = MLPRegressor(max_iter=500)
MLP.fit(sample_train,label_train)
MLP_predict = MLP.predict(sample_test)
accuracy_MLP = r2_score(label_test,MLP_predict)
accuracy_MLP

0.3349344569585142

The result of both Gradient Boost and Neural Network are not have a better performance than the Random Forest, but when Random Forest has better accuracy, it implementing very slow. 